vowpalwabbit/core/include/vw/core/learner.h

```diff
@@ -1054,206 +665,361 @@ class common_learner_builder
   }
 };
 
-template <class DataT, class ExampleT, class BaseLearnerT>
+template <class DataT, class ExampleT>
 class reduction_learner_builder
-    : public common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>, DataT, ExampleT,
-          BaseLearnerT>
+    : public common_learner_builder<reduction_learner_builder<DataT, ExampleT>, DataT, ExampleT>
 {
 public:
-  using super =
-      common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>, DataT, ExampleT, BaseLearnerT>;
-  reduction_learner_builder(std::unique_ptr<DataT>&& data, BaseLearnerT* base, const std::string& name)
-      // NOTE: This is a copy of the base! The purpose is to copy all of the
+  using super = common_learner_builder<reduction_learner_builder<DataT, ExampleT>, DataT, ExampleT>;
+  reduction_learner_builder(std::unique_ptr<DataT>&& data, learner* prev, const std::string& name)
+      // NOTE: This is a copy of the previous learner! The purpose is to copy all of the
       // function data objects so that if this reduction does not define a function such as
       // save_load then calling save_load on this object will essentially result in forwarding the
       // call the next reduction that actually implements it.
-      : common_learner_builder<reduction_learner_builder<DataT, ExampleT, BaseLearnerT>, DataT, ExampleT, BaseLearnerT>(
-            new learner<DataT, ExampleT>(*reinterpret_cast<learner<DataT, ExampleT>*>(base)), std::move(data), name)
-  {
-    // We explicitly overwrite the copy of the base's _finish_example_fd. This
-    // is to allow us to determine if the current reduction implements finish
-    // and in what way.
-    this->learner_ptr->_finish_example_fd = details::finish_example_data{};
-    this->learner_ptr->_learn_fd.base = make_base(*base);
-    this->learner_ptr->_learn_fd.data = this->learner_ptr->_learner_data.get();
-    this->learner_ptr->_sensitivity_fd.sensitivity_f =
-        static_cast<details::sensitivity_data::fn>(details::recur_sensitivity);
-    this->learner_ptr->_finisher_fd.data = this->learner_ptr->_learner_data.get();
-    this->learner_ptr->_finisher_fd.base = make_base(*base);
-    this->learner_ptr->_finisher_fd.func = static_cast<details::func_data::fn>(details::noop);
-    this->learner_ptr->_learn_fd.multipredict_f = nullptr;
-    // Don't propagate merge functions
-    this->learner_ptr->_merge_fn = nullptr;
-    this->learner_ptr->_merge_with_all_fn = nullptr;
-    this->learner_ptr->_add_fn = nullptr;
-    this->learner_ptr->_add_with_all_fn = nullptr;
-    this->learner_ptr->_subtract_fn = nullptr;
-    this->learner_ptr->_subtract_with_all_fn = nullptr;
+      : common_learner_builder<reduction_learner_builder<DataT, ExampleT>, DataT, ExampleT>(
+            prev->make_next_learner(), std::move(data), name)
```

Comment:
I don't love the rename to `prev`, I prefer the original naming of base. Base has also been in use for quite some time in this codebase so I think most people understand what it means

Reply:
Rajan wanted this change because of the double meaning of base - as both a "base learner" that's the last learner in the stack, and a "base of a learner" meaning the learner created immediately before this learner.

I renamed the second use case as "prev" to avoid the overloaded term. I'll see if I can add a comment that clarifies this better.

Reply:
That is a reasonable explanation for splitting them, im on board. I still find the work prev confusing as normally you consider examples traveling down the stack as the default direction. Saying prev in that context implies going up and not down. So it feels like prev is a misnomer

Reply:
I think you're looking at it from the perspective of examples, and I'm describing it from the perspective of learner objects. We create reduction learners by copy-constructing a previously created learner. So each learner has a pointer to its previous, which was the one used to create it.

Reply:
Yeah there are a few ways of looking at it. You should ask a few people to see what the default intuition would be. For me it is the perspective of examples since thats what the logic of reductions is designed around